### ISSUE TITLE
How to get the value of odometry evaluation

### ISSUE THREAD

**Issue by xuwanpeng12** [sconlyshootery/FeatDepth#24]
Hi:
I am sorry to trouble you!
I used the "our model train on kitti odometry" model and “eval_pose.py” provided by you to get the following values without any changes in the parameters.
odom_9 Trajectory error: 0.025, std: 0.010
odom_10 Trajectory error: 0.020, std: 0.011
This seems not as good as "Seq_09  0.016 (std. 0.009)   Seq_10    0.013 (std. 0.009)" of "sfmlearn" .
How did you get "Ours 8.75 2.11 10.67 4.91" and " 3.07 0.89 3.83 1.78"in the paper？
I follow your evaluation code, the code is as follows

eval_pose.py

    from __future__ import absolute_import, division, print_function
    import os
    import sys
    import numpy as np

    import torch
    from torch.utils.data import DataLoader

    sys.path.append('.')
    from mono.datasets.utils import readlines, dump_xyz, compute_ate, transformation_from_parameters
    from mono.datasets.kitti_dataset import KITTIOdomDataset
    from mono.model.mono_fm.pose_encoder import PoseEncoder
    from mono.model.mono_fm.pose_decoder import PoseDecoder

    def evaluate(data_path,model_path,sequence_id,height,width):
        filenames = readlines("/userhome/Feat/FeatDepth3/mono/datasets/splits/odom/test_files_{:02d}.txt".format(sequence_id))

        dataset = KITTIOdomDataset(data_path,
                               filenames,
                               height,
                               width,
                               [0, 1],
                               is_train=False,
                               img_ext='.png',
                               gt_depth_path=None)

        dataloader = DataLoader(dataset,
                            1,
                            shuffle=False,
                            num_workers=4,
                            pin_memory=True,
                            drop_last=False)

        pose_encoder = PoseEncoder(18, None, 2)
        pose_decoder = PoseDecoder(pose_encoder.num_ch_enc)

        checkpoint = torch.load(model_path)
        for name, param in pose_encoder.state_dict().items():
            pose_encoder.state_dict()[name].copy_(checkpoint['state_dict']['PoseEncoder.' + name])
        for name, param in pose_decoder.state_dict().items():
            pose_decoder.state_dict()[name].copy_(checkpoint['state_dict']['PoseDecoder.' + name])
        pose_encoder.cuda()
        pose_encoder.eval()
        pose_decoder.cuda()
        pose_decoder.eval()

        pred_poses = []

        print("-> Computing pose predictions")
        with torch.no_grad():
            for inputs in dataloader:
                for key, ipt in inputs.items():
                    inputs[key] = ipt.cuda()
                all_color_aug = torch.cat([inputs[("color_aug", i, 0)] for i in [0, 1]], 1)
                features = pose_encoder(all_color_aug)
                axisangle, translation = pose_decoder(features)
                pred_poses.append(transformation_from_parameters(axisangle[:, 0], translation[:, 0]).cpu().numpy())
        pred_poses = np.concatenate(pred_poses)

        gt_poses_path = os.path.join(data_path, "poses", "{:02d}.txt".format(sequence_id))
        gt_global_poses = np.loadtxt(gt_poses_path).reshape(-1, 3, 4)
        gt_global_poses = np.concatenate((gt_global_poses, np.zeros((gt_global_poses.shape[0], 1, 4))), 1)
        gt_global_poses[:, 3, 3] = 1
        gt_xyzs = gt_global_poses[:, :3, 3]
        gt_local_poses = []
        for i in range(1, len(gt_global_poses)):
            gt_local_poses.append(np.linalg.inv(np.dot(np.linalg.inv(gt_global_poses[i - 1]), gt_global_poses[i])))

        ates = []
        num_frames = gt_xyzs.shape[0]
        track_length = 5
        for i in range(0, num_frames - 1):
            local_xyzs = np.array(dump_xyz(pred_poses[i:i + track_length - 1]))
            gt_local_xyzs = np.array(dump_xyz(gt_local_poses[i:i + track_length - 1]))
            ates.append(compute_ate(gt_local_xyzs, local_xyzs))

        print("\n  odom_{} Trajectory error: {:0.3f}, std: {:0.3f}\n".format(sequence_id, np.mean(ates), np.std(ates)))

        # save_path = os.path.join(load_weights_folder, "poses.npy")
        # np.save(save_path, pred_poses)
        # print("-> Predictions saved to", save_path)

    if __name__ == "__main__":
        data_path='/userhome/dataset/odometry_color'#path to kitti odometry
        model_path = '/userhome/Feat/FeatDepth3/fm_depth_odom.pth'
        height=320
        width=1024
        sequence_id =9
        evaluate(data_path,model_path,sequence_id,height,width)
        sequence_id = 10
        evaluate(data_path,model_path,sequence_id,height,width)

**Comment by sconlyshootery** [comment]
You can use draw_odometry in script to get the numbers, the metric in eval_pose.py was used in early papers, and most recent papers use the metric in the draw_odometry.py which will import a kitti tool to calculate the numbers.

**Comment by xuwanpeng12** [comment]
Thank you very much！ Following your suggestions, I solved the above problem.
There is another trouble.
when run "cfg_kitti_fm.py   frame_ids =[0,-1,1]", without changing any parameters, I got the following results in epoch 14.
    abs_rel: 0.1046, sq_rel: 0.7124, rmse: 4.5676, rmse_log: 0.1837, a1: 0.8844, a2: 0.9632, a3: 0.9832
The result reached convergence in the 14th epoch and then began to oscillate. This result is not as good as mentioned in your paper. Does this mean that the learning rate or other parameters are set incorrectly? When running “frame_ids =[0,-1,1]” is it different from “frame_ids = [0,-1,1,s]”?

cfg_kitti_fm.py

    DEPTH_LAYERS = 50#resnet50
    POSE_LAYERS = 18#resnet18
    FRAME_IDS = [0, -1, 1]#0 refers to current frame, -1 and 1 refer to temperally adjacent frames, 's' refers to stereo adjacent frame.
    IMGS_PER_GPU = 6 #the number of images fed to each GPU
    HEIGHT = 320#input image height
    WIDTH = 1024#input image width

    data = dict(
        name = 'kitti',#dataset name
        split = 'exp',#training split name
        height = HEIGHT,
        width = WIDTH,
        frame_ids = FRAME_IDS,
        in_path = '/userhome/dataset/kitti',#path to raw data
        gt_depth_path = '/userhome/Feat/FeatDepth0/gt_depths.npz',#path to gt data
        png = False,#image format
        stereo_scale = True if 's' in FRAME_IDS else False,
    )

    model = dict(
        name = 'mono_fm',# select a model by name
        depth_num_layers = DEPTH_LAYERS,
        pose_num_layers = POSE_LAYERS,
        frame_ids = FRAME_IDS,
        imgs_per_gpu = IMGS_PER_GPU,
        height = HEIGHT,
        width = WIDTH,
        scales = [0, 1, 2, 3],# output different scales of depth maps
        min_depth = 0.1, # minimum of predicted depth value
        max_depth = 100.0, # maximum of predicted depth value
        depth_pretrained_path = '/userhome/Feat/FeatDepth0/resnet{}.pth'.format(DEPTH_LAYERS),# pretrained weights for resnet
        pose_pretrained_path =  '/userhome/Feat/FeatDepth0/resnet{}.pth'.format(POSE_LAYERS),# pretrained weights for resnet
        extractor_pretrained_path = '/userhome/Feat/FeatDepth0/autoencoder.pth',# pretrained weights for autoencoder
        automask = False if 's' in FRAME_IDS else True,
        disp_norm = False if 's' in FRAME_IDS else True,
        perception_weight = 1e-3,
        smoothness_weight = 1e-3,
    )

    # resume_from = '/node01_data5/monodepth2-test/model/ms/ms.pth'#directly start training from provide weights
    resume_from = None
    finetune = None
    total_epochs = 40
    imgs_per_gpu = IMGS_PER_GPU
    learning_rate = 1e-4
    workers_per_gpu = 4
    validate = True

    optimizer = dict(type='Adam', lr=learning_rate, weight_decay=0)
    optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
    lr_config = dict(
        policy='step',
        warmup='linear',
        warmup_iters=500,
        warmup_ratio=1.0 / 3,
        step=[20,30],
        gamma=0.5,
    )

    checkpoint_config = dict(interval=1)
    log_config = dict(interval=50,
                      hooks=[dict(type='TextLoggerHook'),])
    dist_params = dict(backend='nccl')
    log_level = 'INFO'
    load_from = None
    workflow = [('train', 1)]

**Comment by sconlyshootery** [comment]
The setting of frame_id =[0,-1,1] and frame_id=[0,-1,1,s] are very different, since 's' will introduce the image in the opposite of the stereo pair, the scale of the depth prediction is fixed.
When s is remove, the scale is ambiguous, 'scale alignment' is performed before evaluation.
I advice you to read [1][2] to know the difference between those two setttings.
[1] Unsupervised Learning of Depth and Ego-Motion from Video
[2] UnsupervisedMonocular Depth Estimation with Left-Right Consistency

**Comment by xuwanpeng12** [comment]
Sorry，maybe my English is too poor!
I have read the article you recommended many times, and I know the difference between inputting stereo pairs and video sequences. I just want to know how do I set the parameters when I train the video sequence as input (frame_id =[0,-1,1]). You do not give specific training parameters using video sequence as input.
Except for "stereo_scale = False, automask = True, disp_norm = True," other parameters remain unchanged, I do not get the results mentioned in your article using video sequences as input(frame_id =[0,-1,1]), do other parameters also need to be changed? Such as learning rate, number of iterations, etc.

**Comment by sconlyshootery** [comment]
Do you want to get the performance on the kitti odometry? I see that your first question is about the evaluation on kitti odometry. If does, you should use the cfg_odom_fm.py for training and the model pretrianed weight is named fm_depth_odom.pth, since the evaluations of depth and pose are conducted in different dataset, so their settings are a bit different.

**Comment by sconlyshootery** [comment]
> Sorry，maybe my English is too poor!
> I have read the article you recommended many times, and I know the difference between inputting stereo pairs and video sequences. I just want to know how do I set the parameters when I train the video sequence as input (frame_id =[0,-1,1]). You do not give specific training parameters using video sequence as input.
> Except for "stereo_scale = False, automask = True, disp_norm = True," other parameters remain unchanged, I do not get the results mentioned in your article using video sequences as input(frame_id =[0,-1,1]), do other parameters also need to be changed? Such as learning rate, number of iterations, etc.

And those settings are correct.

**Comment by sconlyshootery** [comment]
Do you want to get the performance on the kitti odometry? I see that your first question is about the evaluation on kitti odometry. If does, you should use the cfg_odom_fm.py for training and the model pretrianed weight is named fm_depth_odom.pth, since the evaluations of depth and pose are conducted in different dataset, so their settings are a bit different.

**Comment by sconlyshootery** [comment]
I will close this issue, if you have further question, you can reopen it.
